### PR TITLE
Implement degradation comparisons

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Issue #6

## Changes/Additions
1. Shortlisted and added parameters to vary (They are also updated with a Y and an X in the weekly meeting doc)
2. Created `DegradationComparisonGenerator` class and edited + moved `generate_summary_variables` and `experiment_solver` in the class
3. `degradation_parameter_generator` generates a random parameter and a random value.
4. Updated `config_generator` for degradation comparison.

### CODE REVIEW

**Comment by Saransh-cpp** [comment]
Example 1 - 
Sometimes the experiments cut off at different cycles for different values of a parameter, what should be done in this case?

Config - 
```terminal
Single Particle Model CasADi solver with 'safe' mode {'cell geometry': 'arbitrary', 'convection': 'none', 'current collector': 'uniform', 'dimensionality': 0, 'electrolyte conductivity': 'default', 'external submodels': [], 'hydrolysis': 'false', 'lithium plating': 'none', 'lithium plating porosity change': 'false', 'loss of active material': 'none', 'operating mode': 'current', 'particle': 'Fickian diffusion', 'particle mechanics': 'none', 'particle shape': 'spherical', 'SEI': 'ec reaction limited', 'SEI porosity change': 
'true', 'surface form': 'false', 'thermal': 'isothermal', 'total interfacial current density as a state': 'true', 'working electrode': 'none', 'SEI film 
resistance': 'distributed'} Chen2020
[('Discharge at 1 C until 3.5 V', 'Rest for 3 minutes', 'Charge at 1 C until 4.0 V', 'Hold at 4.0 V until 4 mA')]
```

image

**Comment by Saransh-cpp** [comment]
Example 2 - 
For `Ai2020`, 3 subplots are always overlapping

config -
```terminal
Doyle-Fuller-Newman model CasADi solver with 'safe' mode {'cell geometry': 'arbitrary', 'convection': 'none', 'current collector': 'uniform', 'dimensionality': 0, 'electrolyte conductivity': 'default', 'external submodels': [], 'hydrolysis': 'false', 'lithium plating': 'none', 'lithium plating porosity change': 'false', 'loss of active material': 'none', 'operating mode': 'current', 'particle': 'Fickian diffusion', 'particle mechanics': 'swelling and cracking', 'particle shape': 'spherical', 'SEI': 'none', 'SEI porosity change': 'false', 'surface form': 'false', 'thermal': 'isothermal', 'total interfacial current density as a state': 'false', 'working electrode': 'none', 'SEI film resistance': 'none'} Ai2019
[('Discharge at 2 C until 3.6 V', 'Rest for 9 minutes', 'Charge at 3 C until 4.1 V', 'Hold at 4.1 V until 14 mA')]
```

image

**Comment by Saransh-cpp** [comment]
Example 3 -
And then there are some plots which I think are the ones which are ideal -

Config - 
```terminal
Single Particle Model with electrolyte CasADi solver with 'safe' mode 
{'cell geometry': 'arbitrary', 'convection': 'none', 'current collector': 'uniform', 'dimensionality': 0, 'electrolyte conductivity': 'default', 'external submodels': [], 'hydrolysis': 'false', 'lithium plating': 'none', 'lithium plating porosity change': 'false', 'loss of active material': 'none', 'operating mode': 'current', 'particle': 'Fickian diffusion', 'particle mechanics': 'none', 'particle shape': 'spherical', 'SEI': 'solvent-diffusion limited', 
'SEI porosity change': 'true', 'surface form': 'false', 'thermal': 'isothermal', 'total interfacial current density as a state': 'true', 'working electrode': 'none', 'SEI film resistance': 'distributed'} Marquis2019
[('Discharge at 2 C until 3.6 V', 'Rest for 1 minutes', 'Charge at 2 C until 4.0 V', 'Hold at 4.0 V until 49 mA', 'Rest for 10 minutes')]
```

image

**Comment by tinosulzer** [comment]
For example 1 is it saying anything about why the experiments are ending? Could be a solver error causing the red one to end early, 155 Kelvin is crazy low

**Comment by Saransh-cpp** [comment]
> For example 1 is it saying anything about why the experiments are ending? Could be a solver error causing the red one to end early, 155 Kelvin is crazy low

Oh yes, forgot to add a check here, more reasons to work on #33. Unfortunately, I didn't copy the error but there was one.

**Comment by Saransh-cpp** [comment]
Example 4 -

A weird-looking plot

image

File: bot/plotting/random_plot_generator.py [hunk 1]
```diff
@@ -213,6 +213,7 @@ def random_plot_generator(
                     degradation_parameter=degradation_parameter
                 )
 
+                labels = sorted(labels, key=lambda x: float(x.split(':')[1]))
```
Comment (rtimms) [inline bot/plotting/random_plot_generator.py:216]: Does this just sort the labels and not the solutions? I.e. are the labels incorrect now?

Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:216]: Yes, the solutions and labels are jumbled up:)

**Comment by Saransh-cpp** [comment]
Don't merge till the degradation mechanisms are fixed

File: bot/plotting/degradation_comparison_generator.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                    self.degradation_parameter
+                    + ": "
+                    + str(self.param_values[i][self.degradation_parameter]),
+                ]
+            )
+        return sim, solutions_and_labels
+
+    def solve(self):
+        """
+        Solves an expeiment with the given configuration.
```
Comment (brosaplanella) [inline bot/plotting/degradation_comparison_generator.py:93]:
```suggestion
        Solves an experiment with the given configuration.
```

**Review by brosaplanella — APPROVED** [review]
Looks good, thanks Saransh!

**Pull request merged**